sharding/proposer/proposer.go

```diff
@@ -0,0 +1,90 @@
+package proposer
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/sharding"
+	"github.com/ethereum/go-ethereum/sharding/node"
+)
+
+// createCollation creates collation base struct with header
+// and body. Header consists of shardID, ChunkRoot, period,
+// proposer addr and signatures. Body contains serialized blob
+// of a collations transactions.
+func createCollation(n node.Node, shardId *big.Int, period *big.Int, txs []*types.Transaction) (*sharding.Collation, error) {
+	// shardId has to be within range
+	if shardId.Cmp(big.NewInt(0)) < 0 || shardId.Cmp(big.NewInt(sharding.ShardCount)) > 0 {
+		return nil, fmt.Errorf("can't create collation for shard %v. Must be between 0 and %v", shardId, sharding.ShardCount)
+	}
+
+	// check with SMC to see if we can add the header.
+	if a, _ := checkHeaderAvailability(n, shardId, period); !a {
+		return nil, fmt.Errorf("can't create collation, collation with same period has already been added")
+	}
+
+	// serialized tx to blob for collation body.
+	blobs, err := sharding.SerializeTxToBlob(txs)
+	if err != nil {
+		return nil, fmt.Errorf("can't create collation, serialization to blob failed: %v", err)
+	}
+
+	// construct the header, leave chunkRoot and signature fields empty, to be filled later.
+	addr := n.Account().Address
+	header := sharding.NewCollationHeader(shardId, nil, period, &addr, nil)
+
+	// construct the body with header, blobs(serialized txs) and txs.
+	collation := sharding.NewCollation(header, blobs, txs)
+	collation.CalculateChunkRoot()
+	sig, err := n.Sign(collation.Header().Hash())
+	if err != nil {
+		return nil, fmt.Errorf("can't create collation, sign collationHeader failed: %v", err)
+	}
+
+	// add proposer signature to collation header.
+	collation.Header().AddSig(sig)
+
+	return collation, nil
+}
+
+// addHeader adds the collation header to the main chain by sending
+// an addHeader transaction to the sharding manager contract.
+// There can only exist one header per period per shard, it's proposer's
+// responsibility to check if a header has been added.
+func addHeader(n node.Node, collation sharding.Collation) error {
+	log.Info("Adding header to SMC")
+
+	txOps, err := n.CreateTXOpts(big.NewInt(0))
+	if err != nil {
+		return fmt.Errorf("unable to initiate add header transaction: %v", err)
+	}
+
+	var chunkRoot [32]byte
+	copy(chunkRoot[:], collation.Header().ChunkRoot().Bytes())
```

Comment:
Why do you need to make a copy here? Are you mutating it?

Reply:
Not mutating it. I want to convert []byte (what .Bytes() returns) to [32]byte (what SMC takes). Any better idea on how to do that?

Reply:
Does it work if you do this?
```golang
chunkRoot := collation.Header().ChunkRoot().Bytes()[:32]
```

This should reference the same array in-memory, so you aren't copying the data.

Reply:
That doesn't work. I really don't want to copy data. I haven't found a nice way to do it without copying

Reply:
What doesn't work about it? This should be a slice referencing the same internal array.

Reply:
It's still `[]byte` not `[32]byte`. Is this what you are suggesting?
```go
array :=[]byte{0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,31}
array32 := array[:32]
fmt.Printf("%T", array32)
```

Reply:
After more research, the Go language spec does not allow you to access "directly" the underlying, backing array of a slice. We can do that using package [unsafe](https://golang.org/pkg/unsafe/), but as its name hints: it's unsafe

```go
var (
   s []byte    // slice
   a *[32]byte // pointer to array
)

s = make([]byte, 32)
a = byte32(s)
func byte32(s []byte) (a *[32]byte) {
    if len(a) <= len(s) {
        a = (*[len(a)]byte)(unsafe.Pointer(&s[0]))
    }
    return a
}
```